File: src/stores/tasks.ts
import { defineStore } from "pinia";
import { ref, computed } from "vue";
import { api } from "@/composables/useTauri";
import type { Task } from "@/types";

export const useTasksStore = defineStore("tasks", () => {
  const tasks = ref<Task[]>([]);

  const todoTasks = computed(() =>
    tasks.value.filter((t) => t.status === "todo")
  );

  const doneTasks = computed(() =>
    tasks.value.filter((t) => t.status === "done")
  );

  async function fetchTasks(projectId: string) {
    tasks.value = await api.listTasks(projectId);
  }

  async function createTask(projectId: string, title: string) {
    const task = await api.createTask(projectId, title);
    tasks.value.push(task);
    return task;
  }

  async function deleteTask(taskId: string) {
    await api.deleteTask(taskId);
    tasks.value = tasks.value.filter((t) => t.id !== taskId);
  }

  async function updateTaskStatus(taskId: string, status: string) {
    const updated = await api.updateTaskStatus(taskId, status);
    const index = tasks.value.findIndex((t) => t.id === taskId);
    if (index !== -1) {
      tasks.value[index] = updated;
    }
  }

  function clear() {
    tasks.value = [];
  }

  return {
    tasks,
    todoTasks,
    doneTasks,
    fetchTasks,
    createTask,
    deleteTask,
    updateTaskStatus,
    clear,
  };
});

File: src-tauri/src/commands/tasks.rs
use tauri::State;

use crate::models::Task;
use crate::state::AppStateMutex;

use super::projects::save_state;

#[tauri::command]
pub fn create_task(
    state: State<'_, AppStateMutex>,
    project_id: String,
    title: String,
) -> Result<Task, String> {
    let mut state = state.lock().map_err(|e| e.to_string())?;

    let task = Task {
        id: uuid::Uuid::new_v4().to_string(),
        project_id,
        title,
        status: "todo".into(),
        created_at: chrono::Utc::now().to_rfc3339(),
        completed_at: None,
    };

    state.tasks.push(task.clone());
    save_state(&state)?;

    Ok(task)
}

#[tauri::command]
pub fn delete_task(
    state: State<'_, AppStateMutex>,
    task_id: String,
) -> Result<(), String> {
    let mut state = state.lock().map_err(|e| e.to_string())?;
    state.tasks.retain(|t| t.id != task_id);
    save_state(&state)?;
    Ok(())
}

#[tauri::command]
pub fn update_task_status(
    state: State<'_, AppStateMutex>,
    task_id: String,
    status: String,
) -> Result<Task, String> {
    let mut state = state.lock().map_err(|e| e.to_string())?;

    let task = state
        .tasks
        .iter_mut()
        .find(|t| t.id == task_id)
        .ok_or("Task not found")?;

    task.status = status.clone();
    task.completed_at = if status == "done" {
        Some(chrono::Utc::now().to_rfc3339())
    } else {
        None
    };

    let task = task.clone();
    save_state(&state)?;

    Ok(task)
}

#[tauri::command]
pub fn list_tasks(
    state: State<'_, AppStateMutex>,
    project_id: String,
) -> Result<Vec<Task>, String> {
    let state = state.lock().map_err(|e| e.to_string())?;
    Ok(state
        .tasks
        .iter()
        .filter(|t| t.project_id == project_id)
        .cloned()
        .collect())
}
